src/braket/circuits/pauli_string.py

```diff
@@ -0,0 +1,206 @@
+
+        return TensorProduct(
+            [_PAULI_OBSERVABLES[self._nontrivial[qubit]] for qubit in sorted(self._nontrivial)]
+        )
+
+    def weight_n_substrings(self, weight: int) -> Tuple[PauliString, ...]:
+        r"""Returns every substring of this Pauli string with exactly `weight` nontrivial factors.
+
+        The number of substrings is equal to :math:`\binom{n}{w}`, where :math`n` is the number of
+        nontrivial (non-identity) factors in the Pauli string and :math`w` is `weight`.
+
+        Args:
+            weight (int): The number of non-identity factors in the substrings.
+
+        Returns:
+            Tuple[PauliString, ...]: A tuple of weight-n Pauli substrings.
+        """
+        substrings = []
+        for indices in itertools.combinations(self._nontrivial, weight):
+            factors = [
+                self._nontrivial[qubit]
+                if qubit in set(indices).intersection(self._nontrivial)
+                else "I"
+                for qubit in range(self._qubit_count)
+            ]
+            substrings.append(
+                PauliString(f"{PauliString._phase_to_str(self._phase)}{''.join(factors)}")
+            )
+        return tuple(substrings)
+
+    def eigenstate(self, signs: Optional[Union[str, List[int], Tuple[int, ...]]] = None) -> Circuit:
+        """Returns the eigenstate of this Pauli string with the given factor signs.
+
+        The resulting eigenstate has each qubit in the +1 eigenstate of its corresponding signed
+        Pauli operator. For example, a Pauli string +XYZ and signs ++- has factors +X, +Y and -Z,
+        with the corresponding qubits in states |+⟩, |i⟩ and |1⟩ respectively (the global phase of
+        the Pauli string is ignored).
+
+        Args:
+            signs (Union[str, List[int], Tuple[int, ...]], optional): The sign of each factor of the
+                eigenstate, specified either as a string of "+" and "_", or as a list or tuple of
+                +/-1. The length of signs must be equal to the length of the Pauli string. If not
+                specified, it is assumed to be all +. Default: None.
+
+        Returns:
+            Circuit: A circuit that prepares the desired eigenstate of the Pauli string.
+
+        Raises:
+            ValueError: If the length of signs is not equal to that of the Pauli string or the signs
+                are invalid.
+        """
+        qubit_count = self._qubit_count
+        if not signs:
+            signs = "+" * qubit_count
+        if len(signs) != qubit_count:
+            raise ValueError(
+                f"signs must be the same length of the Pauli string ({qubit_count}), "
+                f"but was {len(signs)}"
+            )
+        if isinstance(signs, str) and not set(signs) <= {"+", "-"}:
+            raise ValueError(f"signs must be +/-1, got {signs}")
+        signs_tup = (
+            tuple(_SIGN_MAP[sign] for sign in signs) if isinstance(signs, str) else tuple(signs)
+        )
```

Comment:
I suppose an alternative is
```suggestion
        if isinstance(signs, str):
            if not set(signs) <= {"+", "-"}:
                raise ValueError(f"signs must be +/-1, got {signs}")
            signs_tup = tuple(_SIGN_MAP[sign] for sign in signs)
        else:
            signs_tup = tuple(signs)
```